### ISSUE TITLE
[HOLD for payment 2022-12-20] [$2000] mweb - Copy & Paste - Copy & Pasting text form code editor is not consistent

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#10262]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Access staging.new.expensify.com
2. Sign into a valid account
3. Proceed to any chat
4. Copy text from a code editor and paste it in chat box (Observe the text format)
5. "Copy to clipboard" the text that was just sent and resend it in chat (Observe the code format)

## Expected Result:
The user expects both text formats to be exactly like the one that was copied from the text editor.

## Actual Result:
The two texts differ in format

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Mobile Web

**Version Number:** 1.1.88.1

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** any

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/183101079-56b02fa6-9a66-487b-8ee2-5522f02b28af.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause - Internal Team

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by amyevans** [comment]
Reproducible on mobile web. Good external issue.

**Comment by MitchExpensify** [comment]
[Upwork Job](https://www.upwork.com/jobs/~01bed8d12f86725e61)

**Comment by DharmeshCodeBase** [comment]
Yes i can do this as per your requirement and complete job

**Comment by pecanoro** [comment]
@MitchExpensify Let's double the price

**Comment by ahmdshrif** [comment]
I don't think this is an issue. 

what happens exactly is the screen width is smaller than the line length. 
so the code moved to the next line and that's correct behavior for normal text. 

 but if the user marks it as a code block by using ` ``` `   it will add a scroll block and the long line will not move to the next line. 
 
 Whatsapp has the same behavior : 
 
 
![Screen Shot 2022-08-22 at 4 43 27 PM](https://user-images.githubusercontent.com/21364901/185949730-56a508cd-2d3c-4497-b62a-f6cad053bedd.png)

 
 our behavior when add code block :- 
 
![Screen Shot 2022-08-22 at 4 44 36 PM](https://user-images.githubusercontent.com/21364901/185949856-f2a6054b-5b7b-4994-8c48-3cfcf654720f.png)
 

cc: @parasharrajat , @pecanoro

**Comment by MitchExpensify** [comment]
Doubled!

**Comment by MitchExpensify** [comment]
Doubled!

**Comment by pecanoro** [comment]
So all mWeb platforms will work the same way after this fix? If so, I would say that's fine and to go with the proposal

**Comment by pecanoro** [comment]
@parasharrajat Let me know how you think we should proceed here when you have a moment!

**Comment by parasharrajat** [comment]
Expect an update today.

**Comment by parasharrajat** [comment]
Haven't forgotten this one. I need to review a few things but didn't get the time lately. Prioritizing it. :timer_clock:

**Comment by parasharrajat** [comment]
Update on Friday/Saturday.

**Comment by pecanoro** [comment]
@parasharrajat Any updates?

**Comment by parasharrajat** [comment]
Only need to test the latest proposal on a few platforms to confirm. I will do that asap.

**Comment by parasharrajat** [comment]
Reviewing...

**Comment by parasharrajat** [comment]
There are two issues that are relatable with this GH. 
1. Copying the code content from GH and pasting in our app does not add it as code block. 
2. Copying the formatted string message and then pasting it in the composer does not preserve the spaces. 

I and the proposals above are focusing on 2. 1 is not possible to solve due to the platform does not use `pre` for code. 


For some reason, Returning simple promise works. But it gives error on chrome.
```js
        new ClipboardItem({
            'text/html': Promise.resolve(html),
            ...
        }),
```

@b1tjoy Can you look into the above suggestion? Maybe we don't have to use the sync api/old technique.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by parasharrajat** [comment]
Ok, It is not working. I may be had some cache last time. 


For _Copying the formatted string message and then pasting it in the composer does not preserve the spaces._ https://github.com/Expensify/App/issues/10262#issuecomment-1299362513

 I hate to add checks for safari especially but it seems `navigator.write` API has some issues which we can't control. so @b1tjoy 's [proposal](https://github.com/Expensify/App/issues/10262#issuecomment-1236974635) looks good. The root cause is explained in https://github.com/Expensify/App/issues/10262#issuecomment-1236384132. 
 
cc: @pecanoro 

There are other issues but we decided to drop them in this issue and focus on mWeb. 

:ribbon: :eyes: :ribbon:  C+ reviewed

---

@b1tjoy Are you able to select the text via mouse and copy the text from old messages in the chat?  (not related to this issue)

**Comment by pecanoro** [comment]
Before I approve the proposal, it seems the code that we are adding is only for safari, right? I was able to reproduce this on the browser for android though.

**Comment by parasharrajat** [comment]
Give me sometime and I will check it.

**Comment by parasharrajat** [comment]
Its working for me. Android chrome.

**Comment by pecanoro** [comment]
What? Insane hahaha @MitchExpensify Can you see if you can reproduce it in a mobile browser?

**Comment by parasharrajat** [comment]
Here is a video. 

https://user-images.githubusercontent.com/24370807/199837611-47a13424-b248-4202-ae32-1157d286f626.mp4

**Comment by MitchExpensify** [comment]
On iOS Chrome I copy and pasted:

<ComposeProviders
            components={[
                OnyxProvider,
                SafeAreaProvider,
                SafeArea,
                LocaleContextProvider,
                HTMLEngineProvider,
            ]}
        >
        
Which ended up looking like this:
![IMG_B3C867555965-1](https://user-images.githubusercontent.com/36425901/200025335-1bb911ed-ee96-4069-ba9c-cb626fd819b6.jpeg)

**Comment by parasharrajat** [comment]
Ok. Thanks for confirming the safari copy-paste.

**Comment by parasharrajat** [comment]
Bump @pecanoro

**Comment by parasharrajat** [comment]
Interesting, let me check staging on Android chrome.

**Comment by parasharrajat** [comment]
What is IME? 

Good catch about the secure context. I completely missed that.  In short, I would say that `navigator.clipboard.write` does not serve our purpose. Due to sync pasting being deprecated, we are using it. 

I don't like the second option, it is confusing. 

But I am curious why `navigator.clipboard.write` breaks the formatting. Do you any details on this?

**Comment by MitchExpensify** [comment]
Doubled and created a new [Upwork Job](https://www.upwork.com/jobs/~01a97cccfd627cb0ad)

**Comment by MitchExpensify** [comment]
Any estimate on when the PR  will be ready for this? @parasharrajat

**Comment by parasharrajat** [comment]
For that, we need to hire a contributor first. I will review the latest proposals. We found a new issue during last proposal review.

**Comment by JmillsExpensify** [comment]
@pecanoro To help clear out the extensive backlog of /App bugs, we're putting the spotlight on all bugs older than 4 weeks old. To help unblock the roadmap and get our bug pipeline back in equilibrium, can you:

 - Decide whether any proposals currently meet our guidelines and can be approved as-is
 - For any that can't, please take this issue internal and treat it as one of your highest priorities
 - If you have any questions, don't hesitate to start a discussion in #bug-zero

Thanks everyone!

**Comment by parasharrajat** [comment]
Will review the updated proposal in 2 hours.

**Comment by dylanexpensify** [comment]
@parasharrajat can we get an update

**Comment by parasharrajat** [comment]
@mdneyazahmad Good Idea https://github.com/Expensify/App/issues/10262#issuecomment-1311728174. But it requires a bigger discussion and a lot of changes to the front end and back end. Did you get a chance to raise this on slack as a P/S statement?  

I will keep this off-topic unless there is some discussion about it.

**Comment by parasharrajat** [comment]
> But what if we copy a code block and paste it back use Gboard quick paste suggestions, what do we expected? Code block or plain text? If we want a consistent behavior with native platform, code block is what we want to paste back, not plain text.

We don't want the code block to be pasted via Gboard thus it is fine. It is meant to paste text thus I am fine keeping it as it is. 

@pecanoro You can share your reviews on this. If we do this, it means that we will store MD as `text/plain` in the clipboard, it will be different from `paste and match style` on the Desktop. 

Other than that, I am happy with what we have landed in the final proposal. @b1tjoy [proposal](https://github.com/Expensify/App/issues/10262#issuecomment-1311826800) looks good to me. 

cc: @pecanoro 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by aldo-expensify** [comment]
@parasharrajat @b1tjoy about this proposal: https://github.com/Expensify/App/issues/10262#issuecomment-1311826800

I'm working on some related changes here:
- https://github.com/Expensify/expensify-common/pull/474: `ExpensiMark.htmlToText` no longer collapse multiple `<br />` into a single space because this function was being used to derive `message[0].text` from `message[0].html`, and we don't want whitespace collapsing there.

- https://github.com/Expensify/App/pull/12968: `ReportUtils.formatReportLastMessageText` now has the responsibility of collapsing whitespaces.

- https://github.com/Expensify/Web-Expensify/pull/35658 (this PR): The PR's above changed slightly how the whitespace collapsing work, improving it, this PR is making the same change in the API to produce the same results.



I'm wondering, why do we need a new function `htmlToMultiLineText` and why don't we just make `htmlToText` replace `<br />` by `\n`? Then, for the LHN `lastMessageText`, we can leave it up to `ReportUtils.formatReportLastMessageText` to replace all consecutive spacing (including newlines) by single `' '`. What do you think?

**Comment by parasharrajat** [comment]
I am good with the suggestion @aldo-expensify.  I was also skeptical about two functions.

**Comment by aldo-expensify** [comment]
for some reason, I'm not able to reproduce this in dev (App `main` branch), but I'm able to reproduce using staging.new.expensify.com.

I'm copying this code from github:

https://github.com/Expensify/App/blob/9f790f6b1222116e05b7c57b8e7da0005feb16e0/src/pages/home/report/ReportActionItemFragment.js#L106-L111

**Comment by pecanoro** [comment]
@b1tjoy Let us know when you have reviewed @aldo-expensify message and if you can post an updated proposal, that would be great so I can give you the green light

**Comment by aldo-expensify** [comment]
@pecanoro I just finished testing if replacing `<br />` by `\n` when it comes to urban airship notifications breaks something, and it doesn't... IMO it looks better as it is closer to the original message.

This PR replaced `<br />` for whitespaces in urban airship notifications: https://github.com/Expensify/Web-Expensify/pull/31071. If I understand correctly, this was done to replicate the behaviour of replacing `<br />` by whitespaces in the LHN when we show the last message (https://github.com/Expensify/App/issues/2965).
In the context of urban airship notification, I think we should be replacing `<br>` by new lines and not whitespaces, new lines look fine in this type of notifications.

<table>
<tr>
<th>
Message
</th>
<th>
Notification
</th>
</tr>
<tr>
<td>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/87341702/204366073-1ac25ce5-76f9-4d81-86cc-bae740dc21dd.png">
</td>
<td>
<img width="400" src="https://user-images.githubusercontent.com/87341702/204366303-77536507-f8e9-4976-8c78-6482c75eb229.png">
</td>
</tr>
<tr>
<td>
<img width="589" alt="image" src="https://user-images.githubusercontent.com/87341702/204366168-a61e6355-a6c9-40c4-96a0-eaa8e0ad7726.png">
</td>
<td>
<img width="400" src="https://user-images.githubusercontent.com/87341702/204366218-88b9bc06-f537-4387-89e3-6b4db2d40ad8.png">
</td>
</tr>
</table>

So.. I'm going to work now in finishing these PR's: https://github.com/Expensify/App/issues/10262#issuecomment-1325813786. This means for @b1tjoy 's [proposal](https://github.com/Expensify/App/issues/10262#issuecomment-1311826800) that introducing `htmlToMultiLineText` shouldn't be necessary and `htmlToText` should be used instead.

**Comment by parasharrajat** [comment]
Agree, hold this until https://github.com/Expensify/expensify-common/pull/474 is merged. Only then we can pick up from Aldo's work.

**Comment by aldo-expensify** [comment]
Sorry for the slow progress on this, I'll try to get it done today asap so we can un-hold this one.

**Comment by JmillsExpensify** [comment]
Awesome, that's be ideal!

**Comment by aldo-expensify** [comment]
The expensify-common PR is ready for review now: https://github.com/Expensify/expensify-common/pull/474

From looking at @b1tjoy 's [proposals](https://github.com/Expensify/App/issues/10262#issuecomment-1311826800), I think you should be able to start working:

- Solution of problem 2 is not needed anymore because that is what https://github.com/Expensify/expensify-common/pull/474 is about, right?
- For testing, if you need the changes from https://github.com/Expensify/expensify-common/pull/474, I think you can use the commit hash `2ea21fa3a824ee37e06312114f36bfa0c2d251d5` in your package.json to have the updated version

**Comment by pecanoro** [comment]
@b1tjoy Can you post an updated proposal so we are ready when we remove the HOLD from this?

**Comment by parasharrajat** [comment]
@pecanoro Bump :arrow_up:

**Comment by pecanoro** [comment]
We should avoid writing platform-specific code blocks, what would it happen if we use the same span element for all platforms? Would it break any platforms or the behavior would be the same?

**Comment by parasharrajat** [comment]
That span code is deprecated across browsers because it uses `document.execCommand('copy');`. For the web's sake, the browser still supports this feature. But we don't know when they will stop supporting it. 

The new method is `navigator.clipboard.write([` which has the above-said bug. I also urged the same in the discussion but I think it is best to keep its usage minimal.

**Comment by pecanoro** [comment]
@parasharrajat Sorry, I am not sure if I am following what you mean. Are you agreeing or disagreeing with the proposed solution? I feel like the proposal is a bit hacky and we need something that can work fine across browsers unless it's impossible at all.

**Comment by parasharrajat** [comment]
I already left a review https://github.com/Expensify/App/issues/10262#issuecomment-1301832376 based on the discussion. But the final decision is yours.

**Comment by aldo-expensify** [comment]
> We should avoid writing platform-specific code blocks, what would it happen if we use the same span element for all platforms? Would it break any platforms or the behavior would be the same?

Seems like the code `document.createElement('span');` can't be used for native:

<img width="434" alt="image" src="https://user-images.githubusercontent.com/87341702/205710880-f882122d-c474-48cb-ac21-d208d830a43e.png">

**Comment by parasharrajat** [comment]
This issue is about mWeb platforms and the suggested changes are only for web-based platforms.

**Comment by tgolen** [comment]
Hi! I'm hopping in here because @pecanoro was asking for my opinion on the proposal. 

I think I am mostly OK with the proposal that @parasharrajat has approved. The browser support here is really biting us, and I don't see much of a way around it. It also doesn't look like a full platform solution can be created.

I think there is only one thing I would like to change about this proposal. Right now, the `if safari` check is in `index.js` and that will impact both web and desktop. I think it should be moved inside a `index.website.js` file so that it is not included in our desktop package.

Also, is there a different check we can do rather than looking at the user agent? It would be nice if we can detect the problem as it relates to the actual code (like detecting what the browser _actually_ supports). That way, it doesn't matter what browser it is, or when the browser gets updated to use the new API. It's just, if the browser supports the new API, then use that. Else, fallback to old deprecated method.

**Comment by pecanoro** [comment]
Thank you for the detailed explanation! @b1tjoy Let's go with your proposal then!

**Comment by MitchExpensify** [comment]
I had to make a [new Upwork Job](https://www.upwork.com/ab/applicants/1600240109014396928/job-details) for this and I've invited @parasharrajat. @b1tjoy mind making a proposal on the Upwork job as I'm not 100% on your Upwork profile to invite you myself - Thanks!

**Comment by JmillsExpensify** [comment]
We have one more day to approve the PR and qualify for the 50% bonus! Looks like we have some initial comments from @parasharrajat in the PR.

**Comment by MitchExpensify** [comment]
Hired @b1tjoy for eventual Upwork payment!

**Comment by parasharrajat** [comment]
PR is ready for final review and merge.

**Comment by MitchExpensify** [comment]
I'm heading ooo so reassigning the bug label for a new CM!

**Comment by miljakljajic** [comment]
@b1tjoy , paid you in Upwork! 

@parasharrajat - still waiting on you to approve the offer, then I can pay you!

**Comment by miljakljajic** [comment]
Paid!

**Comment by miljakljajic** [comment]
Ah, you're right! @b1tjoy was assigned and the PR was reviewed and merged between Dec 6 and Dec 8. Paying your 50% bonus now. So sorry for missing that!

**Comment by miljakljajic** [comment]
Paid to both of you.

**Comment by pecanoro** [comment]
@parasharrajat Would you say this bug was a regression of implementing this https://github.com/Expensify/App/pull/9880? It was a new feature so I don't think it's actually a regression. Also, I don't think we need to update the PR reviewer list.

@miljakljajic Did you update the regression tests to catch this bug in the future? Though the bug was caught by Applause, so maybe we can a test already for this.

**Comment by pecanoro** [comment]
@miljakljajic Feel free to close this issue when you confirm that actually the regression tests have been updated to cover this! (if necessary).

**Comment by JmillsExpensify** [comment]
@miljakljajic I'd also link the regression test request to this issue as well. Easier to keep track of everything.

**Comment by miljakljajic** [comment]
Thank you guys! There's an existing test for this (which makes sense given that Applause picked it up, as @pecanoro says):

https://expensify.testrail.io/index.php?/cases/view/1971016&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=229062
